metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala

```diff
@@ -1029,20 +1041,56 @@ class MetalsLspService(
                   if item == Messages.AmmoniteJvmParametersChange.restart =>
                 ammonite.reload()
               case _ =>
-                Future.successful(())
+                Future.unit
             }
-        } else {
-          Future.successful(())
-        }
+        } else if (userConfig.javaHome != old.javaHome) {
+          updateBspJavaHome(session)
+        } else Future.unit
       }
-      .getOrElse(Future.successful(()))
+      .getOrElse(Future.unit)
 
     for {
       _ <- slowConnect
       _ <- Future.sequence(List(restartBuildServer, resetDecorations))
     } yield ()
   }
 
+  private def updateBspJavaHome(session: BspSession) = {
+    languageClient
+      .showMessageRequest(Messages.ProjectJavaHomeUpdate.params())
+      .asScala
+      .flatMap {
+        case Messages.ProjectJavaHomeUpdate.restart =>
+          buildTool match {
+            case Some(sbt: SbtBuildTool) if session.main.isSbt =>
+              for {
+                _ <- disconnectOldBuildServer()
+                _ <- sbt.shutdownBspServer(shellRunner)
+                _ <- sbt.generateBspConfig(
+                  folder,
+                  bspConfigGenerator.runUnconditionally(sbt, _),
+                  statusBar,
+                )
+                _ <- autoConnectToBuildServer()
+              } yield ()
+            case Some(mill: MillBuildTool) if session.main.isMill =>
+              for {
+                _ <- mill.generateBspConfig(
+                  folder,
+                  bspConfigGenerator.runUnconditionally(mill, _),
+                  statusBar,
+                )
+                _ <- autoConnectToBuildServer()
+              } yield ()
+            case _ if session.main.isBloop =>
+              slowConnectToBuildServer(forceImport = true)
+            case _ => Future.successful(())
```

Comment:
I think Bazel will just automatically restart itself if JAVA_HOME changes, so maybe we should not show this in that case?

Reply:
I can skip this message for `Bazel` but otherwise this still needs to be implemented and tested for `Bazel` as a follow up. I don't think it would currently know that `JAVA_HOME` changed.

Reply:
I don't think we should actually do anything in Metals, since this should be done via env variable so that Bazel doesn't restart

Reply:
Okay, then I'll emit a warning here for `Bazel` that this setting is not supported and let's add docs on that.